orttraining/orttraining/training_api/module.cc

```diff
@@ -43,13 +44,16 @@ std::unordered_set<const Node*> GetReverseReachableNodes(Graph& inference_graph,
 }
 
 Status RemoveUnusedNodes(Graph& inference_graph, InlinedVector<const NodeArg*>& output_node_args) {
-  auto reachable_nodes = GetReverseReachableNodes(inference_graph, output_node_args);
+  const auto reachable_nodes = GetReverseReachableNodes(inference_graph, output_node_args);
 
   // Get all graph nodes and remove those that are not in the reachable nodes.
   GraphViewer graph_viewer(inference_graph);
-  for (auto& node : graph_viewer.Nodes()) {
-    if (!reachable_nodes.count(&node)) {
-      inference_graph.RemoveNode(node.Index());
+  const auto node_indices = graph_viewer.GetNodesInTopologicalOrder();
+  for (int node_index = node_indices.size()-1; node_index >= 0; --node_index) {
+    auto* node = inference_graph.GetNode(node_index);
+    if (!reachable_nodes.count(node)) {
+      graph_utils::RemoveNodeOutputEdges(inference_graph, *node);
+      inference_graph.RemoveNode(node_index);
```

Comment:
do we need set graph outputs after removing the nodes? possibly a node producing graph output is removed?

Reply:
A node producing graph outputs will not be removed because it should be a part of `reachable_nodes`.
Not sure I understood the question. Will sync with you about this question.